### ISSUE TITLE
ArResolverContextBinder does not work well with HdStorm multi-threaded commit of textures

### ISSUE THREAD

**Issue by MrKepzie** [PixarAnimationStudios/USD#1886]
### Description of Issue

We have a custom ArResolverContextBinder that allows us to use custom texture from memory. 
The ArResolverContextBinder is thread-local and we use it in the thread that calls HdStRenderDelegate::CommitResources
However HdSt_TextureObjectRegistry::Commit enforces multi-threaded load of textures:

```

        if (_isGlfBaseTextureDataThreadSafe) {
            // Loading a texture file of a previously unseen type might
            // require loading a new plugin, so give up the GIL temporarily
            // to the threads loading the images.
            TF_PY_ALLOW_THREADS_IN_SCOPE();

            // Parallel load texture files
            WorkParallelForEach(result.begin(), result.end(),
                                [](const HdStTextureObjectSharedPtr &texture) {
                                    texture->_Load(); });
        }
```

In our custom asset resolver, the _Resolve and _OpenAsset functions may be called by helper worker threads spawned by tbb. However they do not have the context that was bound by ArResolverContextBinder. As a result of this, _GetCurrentContextObject returns null and we are unable to access our custom textures.

Suggested fix: the threaded lambda  should copy the `_PerThreadContextStack _threadContextStack` member of `_DispatchingResolver`

**Comment by spiffmon** [comment]
Hi @MrKepzie ,
    I am not intimately familiar with the hdSt texture-loading pipeline, but my quick read of the code indicates that we _expect_ Ar and resolution to have already been taken care of by a higher software layer by the time we get into the code you indicated, which I reckon is why we are insensitive to the current ArResolverContext.  I'll tag @unhyperbolic for a more informed opinion.

   But you are of course correct that `WorkParallelForEach` will lose the context.  The easiest way to recapture it is pretty straightforward: fetch the current context to a local, first, and bind it to the lambda; then inside the lambda, before calling `texture->_Load()`, first rebind the context.  There is _some_ cost to doing so, which is why we don't do it every time we invoke Work.

  @unhyperbolic , @tcauchois , I _am_ wondering if we do need to be doing this in UsdImagingDelegate's multithreaded dispatch sites?

**Comment by jilliene** [comment]
Filed as internal issue #USD-7400

**Comment by unhyperbolic** [comment]
It does seem necessary to recapture the resolver context lost by `WorkParallelForEach` and, luckily, @spiffmon already described how to do it.

In more detail:

Most of the asset resolution is done by `UsdImagingBuildHdMaterialNetworkFromTerminal` filling in the resolved path of a `SdfAssetPath` which is then picked up by `hdSt/materialNetwork.cpp`.

Note that `hdSt` still needs an `ArGetResolver` to open an asset through `HioImage`. For example, to open a (resolved) path like `/someDir/myFile.usdz[myImage.exr]` for an asset in a usdz file. The actual call to `ArGetResolver().OpenAsset` is in, e.g., `HioOIIO_Image`.

[ Note that udim textures are a little different: the resolution of symlinks is postponed from usd imaging to hdSt texture loading in `hdSt/udimTextureObject.cpp` where there is an additional call to `ArGetResolver().Resolve` ].

I am not familiar with the expectations of how Ar abilities are registered and whether they are supposed to be still available when dropping the resolver context (given that we have tests for textures in usdz files, it seems that we still can open assets in usdz files). But it seems to me to be more correct to recapture the resolver context.

**Comment by spiffmon** [comment]
Thanks for confirming, @unhyperbolic !  I will tag @sunyab , but I don't think we have strictly defined what clients can or cannot refer to a bound ArResolverContext for, but I definitely *would* say that it could reasonably be relevant to an implementation of `ArGetResolver().OpenAsset()`, if that is what @MrKepzie is doing.

**Comment by sunyab** [comment]
`ArResolverContext` was only intended to factor in to path resolution calls and not calls to `OpenAsset` or other API. This is because downstream clients may not have access to the necessary resolver context to bind it during these operations. 

From the docs [here](https://graphics.pixar.com/usd/release/api/ar_page_front.html):

> It's important to note that any information in a resolver context that is used during path resolution
> should be factored into the resulting resolved path. This is because resolved paths may be passed
> to downstream clients or external libraries that will not (or can not) have a context object bound when
> the resolved path is used to open an asset. Put another way: a resolved path should be usable by 
> downstream clients that do not have a resolver context bound.

This could certainly be more explicit about where we do allow the context to be used, and we could probably restructure the code so that an `ArResolver` implementation could only access the context in allowed functions.

I would not want to change this behavior. For USD itself it would mean that a resolver context would have to be passed down to imaging and usdImaging because clients would be expecting that.

**Comment by spiffmon** [comment]
After chatting with @sunyab , I agree that my earlier suggestion should not be followed, here, assuming (as @unhyperbolic confirmed) that all the texture paths have already been resolved at this point.  I also confirmed that usdImaging already **does** bind the UsdStage's resolver context, but only for the purpose of **resolving** texture paths.  @unhyperbolic , if any of that is happening in parallel (e.g. executing UsdImagingMaterialAdapter), we **would** need to propagate the currently-bound context.
@MrKepzie , are you able to follow the prescribed pattern of extracting, during Resolve(), any needed info from the current Context, such that a later call to `OpenAsset` would find everything it needs in the resolved asset path itself?

**Comment by MrKepzie** [comment]
Below a screenshot of the callstack of _Resolve:
![Screenshot 2022-06-01 at 08 07 04](https://user-images.githubusercontent.com/4058448/171338645-50e1afdb-374f-4ec6-93e3-10b33772e8fd.png)

in `UsdRenderEffect::render` we bind our custom `ArResolverContext`. 
However  `_MakeResolvedAssetPathsImpl` later on in the call-stack binds another ArResolverContext returned by `GetPathResolverContext()` (which is the default ArResolver::CreateDefaultContextForAsset).

I'm not sure I understand what you mean we should do during _Resolve, there isn't much we can do as it is already called in a separate thread with the stage's context bound. (see callstack below)

Should we assign our custom ArResolverContext to the stage and  emulate the same funtionnality as the default resolver ?
If so, what's the point of ArResolverContextBinder if only the one passed to the UsdStage is reliable when implementing a custom ArResolver? (I may be missing other use-cases)

![Screenshot 2022-06-01 at 08 23 48](https://user-images.githubusercontent.com/4058448/171340852-6e51f0f1-2274-419a-82ef-5804397c5a77.png)

**Comment by spiffmon** [comment]
Thanks for the callstacks, they clear things up.  The question you ask is a very reasonable one, though of course there are places in the internals (like inside the multithreaded composition engine), where we need ArResolverContextBinder.  I think that the reason we ask the stage to resolve its own assets paths using the context with which the stage was created is so that we get consistent results from the stage no matter where it is queried.  We'll discuss this internally, but one question that may be relevant (since it would be problematic to completely ignore the stage's primary-resolver context) is whether you can deploy your in-memory textures as an add-on URI resolver instead of as a custom _primary_ resolver?  And in the meantime, yes, if you are able to create/open the stage _with_ your custom resolver context, then that seems like the easiest "sanctioned" workaround.

And thank you also for the second stack: as @sunyab said, we do not encourage re-resolving the texture's path from inside an ArAsset implementation.
